### PR TITLE
Change compound column field name rules

### PR DESCRIPTION
**Pull request by waynexia**
# Which issue does this PR close?

Closes #.

 # Rationale for this change

Origin in #792, some discussions here: https://github.com/apache/arrow-datafusion/pull/792#issuecomment-902936271

In the current implementation and spec., all field names don't contain the table or relation name. It might be better to hold those qualifiers in compound columns resulting from projection with calculating expr (`select t1 + 1`) or aggr expr (`sum(t2)`) etc. This can improve readability by figuring out where a column is from. And also in some parts of inner processing (like the logical plan phase), we use this qualified form to identify a column.

# What changes are included in this PR?

This PR changes the rule of field name generation. For now, all compound columns' (columns with other outer expressions) field names will be qualified form.

# Are there any user-facing changes?

Yes. 

Compound columns' field name will now contain table / relation name, like `AVG(table.c1)`.

### CODE REVIEW

**Comment by waynexia** [comment]
Hi @houqp, could you please take a look at this?

And according to the PR template, this PR should be labeled as "API change". But I can't add the label myself.

**Comment by jorgecarleitao** [comment]
Thanks for the proposal.

imo it is a bit counter-intuitive to write `SELECT MIN(c1) FROM table1` and get `MIN(table.c1)` as the schema. Are there other engine doing this?

I am curious how this interacts with the dataframe API. E.g. (pseudo-code)

```rust
df = ctx.create_in_memory_table("table1", batches)?;
df = df.agg(sum(df["c1"]))?;
df.collect()
```

would yield the schema `"SUM(table1.c1)"` or `"SUM(c1)"`? Would I need to write `df = df.agg(sum(df["table.c1"]))?;` instead?

**Comment by waynexia** [comment]
Thanks for your question @jorgecarleitao.

> imo it is a bit counter-intuitive to write SELECT MIN(c1) FROM table1 and get MIN(table.c1) as the schema. Are there other engine doing this?

Not that I know of for now. There are some other engines' behavior [here](https://github.com/apache/arrow-datafusion/blob/master/docs/specification/output-field-name-semantic.md#function-transformed-columns). This proposal does change a lot. From eliminating all qualifiers to bring all to those composite expressions.

> would yield the schema "SUM(table1.c1)" or "SUM(c1)"

It will be `SUM(table1.c1)`. The way of interacting with dataframe doesn't change. Only the result field names are different. There is one [example](https://github.com/apache/arrow-datafusion/pull/952/files#diff-1ca1a4d1e2c60e21a139b211b92f05889ae035c2fa9aa87bc878c32f26606fabR262-R291) in `execution/dataframe_impl.rs`

**Comment by houqp** [comment]
Thanks @waynexia for the PR, I am on vacation this week, but I will try to find some time to review it today or tomorrow.

**Review by alamb — COMMENTED** [review]
I defer to @houqp  and @jorgecarleitao  on this -- they spent quite some time studying and writing up the output field name semantics: https://github.com/apache/arrow-datafusion/blob/master/docs/specification/output-field-name-semantic.md which I think we should continue to follow, if possible

cc @Jimexist

**Comment by houqp** [comment]
Our current output field name semantics mostly aligns with what spark has, which strips column qualifiers in all cases.

This PR changes the semantics to handle compound and bare column names differently. For bare names, we strip column qualifiers, but not for compound columns.

Unlike Spark, relational databases like MySQL, Posgresql and Sqlite all treat compound and bare column names differently. MySQL, Postgresql and Sqlite strip qualifers for bare column names like we do. But MySQL and Sqlite use raw user query input for compound column names. Postgresql on the other hand just uses `?column?` for all compound column names.

In all compute engines, users are not expected to reference compound columns by generated names because these names are not guaranteed to be valid sql expressions. Instead, they should always manually alias them. As a result, the output filed name for compound columns are just there for display/debug purpose only. See also @Dandandan 's comment at https://github.com/apache/arrow-datafusion/pull/280#issuecomment-834805975.

@jorgecarleitao as for the counter-intuitive example you mentioned, the current implementation will output field name `SUM(id)` column name for query `SELECT SUM(t1.id)`, while the proposed new behavior will output `SUM(t1.id)` field name for query `SELECT SUM(id)`. So both of them will not use the exact user query input as the output field name. Either way, it should have no impact to how users construct queries.

The proposed new behavior provides better UX compared to Postgresql's `?column?` column name. I think it's also an improvement over the current behavior because it will produce an unambiguous column name for queries like `SELECT t1.id * t2.id FROM t1 JOIN t2 USING id`. The current behavior will output `id * id`, which is not as good as `t1.id * t2.id` IMO.

File: datafusion/src/physical_plan/planner.rs [hunk 1]
```diff
@@ -76,15 +75,25 @@ fn create_function_physical_name(
     Ok(format!("{}({}{})", fun, distinct_str, names.join(",")))
 }
 
-fn physical_name(e: &Expr, input_schema: &DFSchema) -> Result<String> {
+fn physical_name(e: &Expr) -> Result<String> {
```
Comment (houqp) [inline datafusion/src/physical_plan/planner.rs:78]: nitpick, by just looking at the function names, it's not easy to tell the difference between `physical_name` and `create_physical_name`. Perhaps rename `physical_name` to `root_physical_name`?

Alternatively, we could get rid of the wrapper, then use `physical_name` and explicitly pass in the `is_first_expr` argument everywhere.

**Review by houqp — APPROVED** [review]
LGTM, left two minor naming comments.

**Comment by waynexia** [comment]
Thanks @alamb and @houqp for reviewing! I pushed a commit to address the comments.

File: datafusion/src/physical_plan/planner.rs [hunk 2]
```diff
@@ -76,15 +75,25 @@ fn create_function_physical_name(
...
+    create_physical_name(e, true)
+}
+
+fn create_physical_name(e: &Expr, is_first_expr: bool) -> Result<String> {
```
[the remark this answers is not in the document]
Comment (waynexia) [reply datafusion/src/physical_plan/planner.rs:82]: I picked the first one and explained how it is used in the doc comment

File: datafusion/src/physical_plan/planner.rs (hunk 1 above)
Comment (waynexia) [reply datafusion/src/physical_plan/planner.rs:78]: It's a good point to remove the wrapper. `physical_name()` is an interior helper function and adding that argument won't be too verbose.

**Review by houqp — APPROVED** [review]
LGTM, I will give some time for @alamb  @jorgecarleitao and @Dandandan to review and comment.

**Review by jorgecarleitao — APPROVED** [review]
Thanks a lot for clarifying @waynexia and @houqp .

This looks great, the code and tests look solid, and the changes to the spec also look great.

:shipit:

Thanks for this PR, @waynexia =)

**Comment by alamb** [comment]
Thanks @waynexia !

**Pull request merged**